### PR TITLE
goroutine util

### PR DESCRIPTION
**Pull request by qiuax**
What type of PR is this?
 Add goroutine util:
code in project   /common/goroutine/goroutine.go

What this PR does / why we need it:
The goroutine in the existing code is not robust, and this panic is not captured. If Ctrip internal panic, it will cause the main Ctrip to crash

Which issue(s) this PR fixes:
This PR fixes the capture of panic in goroutine, even if panic appears inside the goroutine, the main Ctrip can operate normally.

Signed-off-by: qiuax
/assign @kevin-wangzefeng

### CODE REVIEW

File: common/goroutine/goroutine.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package goroutine
+
+import (
+	"log"
+	"reflect"
+)
+
+func Goroutine(function interface{}, args ...interface{}) {
```
Comment (chendave) [inline common/goroutine/goroutine.go:8]: Can you explain where this is used? I can not see it in this PR.

File: cloud/pkg/cloudstream/containermetrics_connection.go [hunk 2]
```diff
@@ -74,9 +75,8 @@ func (ms *ContainerMetricsConnection) SendConnection() (stream.EdgedConnection,
 		URL:    *ms.r.Request.URL,
 		Header: ms.r.Request.Header,
 	}
-	targetPort := strings.Split(ms.r.Request.Host, ":")[1]
 	connector.URL.Scheme = httpScheme
-	connector.URL.Host = net.JoinHostPort(defaultServerHost, targetPort)
+	connector.URL.Host = net.JoinHostPort(defaultServerHost, strconv.Itoa(constants.ServerPort))
```
Comment (chendave) [inline cloud/pkg/cloudstream/containermetrics_connection.go:79]: How this matter for the goroutine you proposed?

**Review by chendave — COMMENTED** [review]
> If Ctrip internal panic, it will cause the main Ctrip to crash

And what's the Ctrip means here?

**Comment by qiuax** [comment]
Hello, I resubmitted the pull request, you can take a look and see if it is reasonable?




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Dave (Wei) ***@***.***&gt;; 
发送时间: 2021年9月1日(星期三) 上午10:29
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [kubeedge/kubeedge] goroutine util (#3082)





 
@chendave commented on this pull request.
  
If Ctrip internal panic, it will cause the main Ctrip to crash
  
And what's the Ctrip means here?

 
In common/goroutine/goroutine.go:
 &gt; @@ -0,0 +1,35 @@ +package goroutine + +import ( +	"log" +	"reflect" +) + +func Goroutine(function interface{}, args ...interface{}) {  
Can you explain where this is used? I can not see it in this PR.
 
 
In cloud/pkg/cloudstream/containermetrics_connection.go:
 &gt;  	connector.URL.Scheme = httpScheme -	connector.URL.Host = net.JoinHostPort(defaultServerHost, targetPort) +	connector.URL.Host = net.JoinHostPort(defaultServerHost, strconv.Itoa(constants.ServerPort))  
How this matter for the goroutine you proposed?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by qiuax** [comment]
#3095




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Dave (Wei) ***@***.***&gt;; 
发送时间: 2021年9月1日(星期三) 上午10:29
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [kubeedge/kubeedge] goroutine util (#3082)





 
@chendave commented on this pull request.
  
If Ctrip internal panic, it will cause the main Ctrip to crash
  
And what's the Ctrip means here?

 
In common/goroutine/goroutine.go:
 &gt; @@ -0,0 +1,35 @@ +package goroutine + +import ( +	"log" +	"reflect" +) + +func Goroutine(function interface{}, args ...interface{}) {  
Can you explain where this is used? I can not see it in this PR.
 
 
In cloud/pkg/cloudstream/containermetrics_connection.go:
 &gt;  	connector.URL.Scheme = httpScheme -	connector.URL.Host = net.JoinHostPort(defaultServerHost, targetPort) +	connector.URL.Host = net.JoinHostPort(defaultServerHost, strconv.Itoa(constants.ServerPort))  
How this matter for the goroutine you proposed?
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by fisherxu** [comment]
> > If Ctrip internal panic, it will cause the main Ctrip to crash
> 
> And what's the Ctrip means here?

goroutine 协程 携程, it should be  协程 goroutine, not Ctrip. @chendave ;-)

**Comment by qiuax** [comment]
ok,util can be used later.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Fisher ***@***.***&gt;; 
发送时间: 2021年9月1日(星期三) 中午11:56
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; "State ***@***.***&gt;; 
主题: Re: [kubeedge/kubeedge] goroutine util (#3082)





   
If Ctrip internal panic, it will cause the main Ctrip to crash
  
And what's the Ctrip means here?
  
goroutine 协程 携程, it should be  协程 goroutine, not Ctrip. @chendave ;-)
 
—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by chendave** [comment]
> goroutine 协程 携程, it should be 协程 goroutine, not Ctrip. @chendave ;-)

it's funny! :)

**Pull request closed without merging**